Add riff-raff task to deploy static assets

MatthewJWalls opened the pull request:
## What does this change?

implement deploying and sourcing assets.

Changes:

* Rewrote sh script for creating the riffraff bundle into js
* Renamed the make tasks to riffraff-bundle and riffraff-publish
* RiffRaff configuration to upload static assets to our assets.guim domain
* Altered asset-paths to use assets.guim in production mode and dev in dev mode

## Why?

Enable assets deploy

MatthewJWalls commented:
Works as expected! 

![image](https://user-images.githubusercontent.com/1218561/37606331-276a8f1e-2b8d-11e8-9503-9fb0e2e8cdcf.png)

Now to make it do the real files, if they're available

MatthewJWalls commented:
I think we need to implement/configure something on the js side that separates our dist directory into javascripts/stylesheets/etc so that the deploy script doesn't have that responsibility. Not sure how difficult that will be

sndrs commented on __tasks__/build-riffraff-artifact.sh:
```diff
@@ -17,14 +17,18 @@ if [[ -z $BRANCH_NAME ]]; then
 
 zip -r guui.zip * -x node_modules/**\*
 [ -d target ] && rm -rf target
+mkdir -p $ASSETS_ROOT/moon-javascripts
```
i feel like we're going to need to settle on one name soon! `moon` in repo called `guui`...

MatthewJWalls commented:
Yeah it's a stub so not final naming. I do think we should rename guui to
remote-render or something though at some point

On 20 Mar 2018 10:23, "Alex Sanders" <notifications@github.com> wrote:

> *@sndrs* commented on this pull request.
> ------------------------------
>
> In __tasks__/build-riffraff-artifact.sh
> <https://github.com/guardian/guui/pull/64#discussion_r175715085>:
>
> >
>  zip -r guui.zip * -x node_modules/**\*
>  [ -d target ] && rm -rf target
> +mkdir -p $ASSETS_ROOT/moon-javascripts
>
> i feel like we're going to need to settle on one name soon! moon in repo
> called guui...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/guardian/guui/pull/64#pullrequestreview-105287691>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKYAXkZhjTUioGxMyqDAhwNbxcBLCp2ks5tgNibgaJpZM4SwaCr>
> .
>

-- 

------------------------------
This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

MatthewJWalls commented:
@sndrs @SiAdcock 

Ok.

I've stripped down the old .sh assets script into basically nothing, and rewrote it in JS, and added some code to copy assets into the bundle. The only thing the bash script does now is run the s3 upload but that could be rewritten into js as well.

* I haven't used much of javascript's promise/async stuff before so the way I've organised the script might be rubbish, let me know
* I think a lot of the work this script is doing might be able to be something webpack can be configured to do, but I'm not sure - at least in reading this script you can get an idea of what I'm attempting to achieve and then let me know how to do it properly!

sndrs commented on __tasks__/build-riffraff-bundle.js:
```diff
@@ -0,0 +1,104 @@
+const fs = require('fs');
+const { promisify } = require('util');
+const { warn, log } = require('./log');
+const exec = promisify(require('child_process').execFile);
```
we use https://github.com/sindresorhus/execa elsewhere, which near enough gives you this – should be a quick drop in i think?

MatthewJWalls replied:
+1

sndrs commented on __tasks__/build-riffraff-bundle.js:
```diff
@@ -0,0 +1,104 @@
+
+function copyJavascript(root = 'target') {
+    return cpy(
+        ['dist/*.js', 'dist/manifest.json'],
```
i feel like needing to know how the riffraff deploy handles different file types that come out of bundling is going to bite us at some point. do we need to break them up (rather than just move all of dist)?

MatthewJWalls replied:
How we arrange the files is entirely up to us really. Rifraff just takes our directory structure underneath target/frontend-static and transfers it to s3 where it's available on the assets.guim.co.uk domain. But we can arrange/name them however we want

MatthewJWalls replied:
I guess the organisation under dist is something webpack handles - so if we can have webpack deciding how to organise dist, and have the script just copy all of dist, that could work too. We still have the static fonts css file which is only outside of dist as well though.

MatthewJWalls commented:
Does anybody know what I can do about the travis lint error?

```
/home/travis/build/guardian/guui/__tasks__/build-riffraff-bundle.js
  3:15  error  'execa' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies
```

I'm not sure moving execa out of devDependencies is valid since it's only used in this task?

SiAdcock commented:
@MatthewJWalls `dependencies` are things that the application needs to run, whereas `devDependencies` are only needed by application developers. I'd argue that `execa` is needed to run the application, because without it, the application won't even get deployed. Arguably by this definition, this means a lot of our `devDependencies` should actually be `dependencies`.

Personally, I think the `dependencies` vs `devDependencies` is more relevant to libraries. When you install a library in your application, you don't want all of its `devDependencies` to become dependencies of your application.

**tl;dr** I'd suggest moving `execa` to `dependencies`

sndrs commented on src/__lib__/asset-path.js:
```diff
@@ -9,4 +9,12 @@ try {
-export default name => `/assets/javascript/${assetHash[name] || name}`;
+const prefix = (name) => {
+    if(process.env.NODE_ENV === "production"){
+        return `https:\/\/assets.guim.co.uk/javascript/${name}`;
```
do you need to escape these?

MatthewJWalls commented:
I'm happy with this now. It all works, deploy has been tested.
